HDXDSYS-561 Add support for env var HDX_KEY_STAGE, HDX_KEY_DEV etc.

mcarans opened the pull request:
Update pre-commit config

Fixes #62

IanHopkinson commented:
I tried cloning the repo to propose a fix, however I get a number of test fails before I touch anything:
1. TestCKAN.test_create_dataset fails because the environment variable GSHEET_AUTH is not defined (as an aside I can't see the environment variables set on this repo). I'm not sure that using a Google Sheet in this situation is necessary - a CSV test fixture(s) would be easier.
2. TestConfiguration.test_env_vars fails because I have HDX_KEY_STAGE defined in my `.hdx_configuration.yaml`
3. TestDatasetCore.test_create_in_hdx and TestDatasetCore.test_update_in_hdx both fail with when trying to do `remove(file.name)` with an error like: >       
```
remove(file.name)
E       PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\ianho\\AppData\\Local\\Temp\\tmptbfxct85' 
```
4. TestDatasetUpdateResourcesLogic.test_dataset_update_resources fails because the paths to the resources are hardcoded assuming UNIX-like paths.


I'm happy to fix these tests for running on Windows but if you don't think that's a priority then please let me know.

mcarans commented:
@IanHopkinson I've sent you GSHEET_AUTH. I used a mixture of Google drive remote urls and filestore uploads to ensure that resources work for both cases.

    TestConfiguration.test_env_vars fails because I have HDX_KEY_STAGE defined in my .hdx_configuration.yaml
I've made some changes and tested defining hdx_key_stage and the tests pass now.

    TestDatasetCore.test_create_in_hdx and TestDatasetCore.test_update_in_hdx both fail with when trying to do remove(file.name) with an error like: >

remove(file.name)
E       PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\ianho\\AppData\\Local\\Temp\\tmptbfxct85' 
I have changed to rely on closing the file deleting the temporary file instead of explicitly deleting it. Hopefully that should fix this problem.

    TestDatasetUpdateResourcesLogic.test_dataset_update_resources fails because the paths to the resources are hardcoded assuming UNIX-like paths.
I've fixed the UNIX paths I found and changed to use a join instead.

If you see any other Windows test failures, please go ahead and fix.

I've made the logic clearer so that HDX_KEY_STAGE is always preferred to HDX_KEY whether in environment variable, keyword argument or configuration. 

The coverage decrease is simply because there are less lines in the Configuration.py because I simplified the logic. I've gone into the coveralls settings and added a coverage decrease threshold which should hopefully prevent this sort of failure in future.

IanHopkinson commented on tests/hdx/api/test_configuration.py:
This test now passes

IanHopkinson commented on src/hdx/api/configuration.py:
revised code looks good it removes the division between environment variable and file parsing for configuration variables

IanHopkinson reviewed (approved):
This looks good - my reading is that the "environment variables" in .hdx_configuration.yaml and real environment variables are handled together in the same function and thus their treatment is identical

I have made a couple of fixes so that all tests run locally on my Windows machine but one test now fails in GitHub Actions...
